torchtune/utils/_distributed.py

```diff
@@ -345,12 +348,45 @@ def get_full_model_state_dict(
     """
     sharded_sd = model.state_dict()
     cpu_state_dict = {}
-    for param_name, sharded_param in sharded_sd.items():
-        full_param = sharded_param.full_tensor()
-        if is_rank_zero:
-            cpu_state_dict[param_name] = full_param.cpu()
-        else:
-            del full_param
+    has_nf4 = any(
+        isinstance(param._local_tensor, NF4Tensor) for param in model.parameters()
+    )
+    if has_nf4:
+        from torch.distributed._composable.fsdp.fully_shard import FSDPModule
+
+        # Iterating from lowerer modules to higher
+        # Unsharding lora adapters before unsharding transformer block
+        for module_name, module in reversed(list(model.named_modules())):
+            if not isinstance(module, FSDPModule):
+                continue
+            module.unshard(async_op=False)
```

Comment:
Is this line where your comment 

> We need to reserve memory for 1 TransformerBlock

is relevant? Also will we ever run into the case where we accidentally unshard the full model?

Reply:
exactly! `module.unshard` is the reason of 1 x TransformerBlock

> accidentally unshard the full model

the unspoken knowledge is `module.unshard` is not recursive. if we call `.unshard` for root model, it will unshard tok_embeddings but NOT TransformerBlock

I also ensure we call unshard from lower module to higher module with `reversed(list(model.named_modules()))`. after `module.unshard(async_op=False)`, we always call `module.reshard()` to ensure freeing memory immediately